Support kill() process when GDB does not exit

bangdang1411 opened the pull request:
Following the discussion in PR #256
I would like to commit the new change about it. I think it is easy to review because the change is different from #256

Solution: Override disconnectRequest() with option exitGDBServer() if this request does not use for restart Request. It has some benefits:
1. disconnectRequest does not wait over 2 seconds for response. exitGDBServer() will handle separately with the request. 
2. restartRequest will be implemented in RenesasGDBTargetDebugSession, that mean it is not depended on timeout of others Request.
3. does not need to implement terminateRequest because it seems cannot update CallStack view when GDB exited.

bangdang1411 commented:
Hi @jonahgraham-san,
I have read this discussion in #256 and #258 PR and would like to commit my solution.
This commit can fix for Renesas devices, but it needs to be reviewed and confirmed for issue #258.
The fix for issue #256 about "restartRequest" must change some things in RenesasGDBTargetDebugSession and override the restartRequest() function.
Thank you.

jonahgraham commented:
If you are going to override `exitGDBServer` in Renesas's backend, why not just override `disconnectRequest` entirely then? Then we can merge #258 for the general case and Renesas can do whatever it wants?

PS. If one of the reasons you want to change it here is for the `isServerTerminated`, then I suggest you poll `this.gdbserver.exitCode`, or better add an additional listener on gdbserver's `'exit'` (and have it set a promise so you can remove polling altogether)

bangdang1411 commented:
Hi @jonahgraham-san,
> If you are going to override exitGDBServer in Renesas's backend, why not just override disconnectRequest entirely then? Then we can merge https://github.com/eclipse-cdt-cloud/cdt-gdb-adapter/pull/258 for the general case and Renesas can do whatever it wants.

My solution is not to override **exitGDBServer()** and **disconnectRequest** in Renesas's backend. I will override **restartRequest**, please see the commit in renesas-gdb-adapter that I comment in the Jira ticket.
This commit in cdt-gdb-adapter will fix the issue of #258, It supports kill gdb process after 4 seconds timeout (it is not dependent on 2 seconds timeout of **disconnectRequest**).
The bug restart Renesas devices will use this commit to implement **restartRequest**.

> PS. If one of the reasons you want to change it here is for the isServerTerminated, then I suggest you poll this.gdbserver.exitCode, or better add an additional listener on gdbserver's 'exit' (and have it set a promise so you can remove polling altogether)

Thank you for your suggestion. Please help me review the latest commit and give me your idea.
Do we need to write the integration test for it? Thank you.

jonahgraham commented on src/GDBTargetDebugSession.ts:
```diff
@@ -235,6 +238,11 @@ export class GDBTargetDebugSession extends GDBDebugSession {
             }
 
             this.gdbserver.on('exit', (code) => {
+                this.isServerTerminated = true;
```
Does my earlier suggestion of `this.gdbserver.exitCode` not work?

bangdang1411 replied:
I think using `isServerTerminated` that can be used in Renesas's backend also.
I do not use `this.gdbserver.exitCode` but I think it is not different.

jonahgraham replied:
But renesas backend can use this.gdbserver.exitCode and then it removes this redundant code.

bangdang1411 replied:
I changed it in the latest commit.

jonahgraham commented on src/GDBTargetDebugSession.ts:
```diff
@@ -341,4 +349,45 @@ export class GDBTargetDebugSession extends GDBDebugSession {
+        try {
+            await this.gdb.sendGDBExit();
+            // Kill a process in case of termination, should not in case of restart
+            if (_args.restart !== true) {
```
I don't think I understand the `restart` flow. Why do you leave the gdb server running if restart is true?

bangdang1411 replied:
In the **restart** flow, I think GDB should be exited normally. Kill process that makes it cannot launch again in case of some hardware device (Renesas RL78, RH850,..)

jonahgraham replied:
I don't really follow, but I assume I will understand once I see the rest.

jonahgraham commented on src/GDBTargetDebugSession.ts:
```diff
@@ -341,4 +349,45 @@ export class GDBTargetDebugSession extends GDBDebugSession {
+            await this.gdb.sendGDBExit();
+            // Kill a process in case of termination, should not in case of restart
+            if (_args.restart !== true) {
+                // Call async function for checking process exit gdb after send response to client
```
Does this work? IIUC VSCode will kill the process tree after the disconnect response? See the VSCode source [here](https://github.com/Microsoft/vscode/blob/e9229a18805fdb9baf7356ddc31bd216da5842ef/src/vs/workbench/contrib/debug/browser/rawDebugSession.ts#L572-L576) and [here](https://github.com/Microsoft/vscode/blob/990a496681948ec274da5503b35cd49a7d06c0a2/src/vs/workbench/contrib/debug/node/debugAdapter.ts#L268-L285)

bangdang1411 replied:
I think yes, the async function monitors the process of GDB in server, I think it is asynchronous with the process tree in VS Code.
Is my understanding correct?

bangdang1411 replied:
>But can you actually see the node process for the adapter running for the 4 seconds after disconnect returns?

I test with RL78 device in some PCs that gdb is killed after 4 seconds and the emulator hangs. It cannot debug again until the voltage in emulator resets.

bangdang1411 replied:
I think I should check the node process again. Maybe I misunderstand in some where.

bangdang1411 replied:
I can confirm the node process still existed after 4 seconds. But I am not sure how `stopSession()` in vs code takes much time to kill this process.

jonahgraham commented on src/GDBTargetDebugSession.ts:
```diff
@@ -341,4 +349,45 @@ export class GDBTargetDebugSession extends GDBDebugSession {
+        if (isKillProcess) {
+            // Kill process after 4 seconds timeout
+            timeout = setTimeout(() => {
+                this.gdbserver?.kill();
```
On Linux this will (only) do a SIGTERM so the gdbserver may still be running which will lead to the await on line 390 to wait forever.

bangdang1411 replied:
Does Linux not have an event `exit` to resolve this promise and we cannot use kill() gdb process? Is my understanding correct?
Can you share the information about it in Linux? 
Sorry I have not understood in Linux environment.

jonahgraham replied:
See this reasonable explanation of sigterm vs sigkill. https://linuxhandbook.com/sigterm-vs-sigkill/

jonahgraham replied:
Anyways, lets put the Linux support to the side for the moment and get the Windows one working properly.

bangdang1411 replied:
Thank you for your information. I will check for both Linux and Window.

bangdang1411 replied:
My Linux environment has some config error, so I cannot test in the case of Linux.
Could you please help me take a look at my code, is send `kill('SIGTERM')` match your idea in Linux?
Or not I will fix my environment tomorrow and retest it again. Thank you.

jonahgraham reviewed (commented):
> Do we need to write the integration test for it? Thank you.

Yes. If it helps, the answer to "do we need a test" is always yes, or an explanation as to why it is difficult or impractical to write an automated test.

bangdang1411 commented on src/GDBTargetDebugSession.ts:
```diff
@@ -341,4 +349,45 @@ export class GDBTargetDebugSession extends GDBDebugSession {
+        }
+    }
+
+    protected async exitGDBServer(isKillProcess: boolean): Promise<void> {
```
I think kill process that should be used in case of **disconnect** or **terminate** the gdb, should not be in case of restart.
If GDB is exited by kill, that means it cannot debug again if user does not reset the voltage in the emulator (Checking by Renesas devices: RL78, RH850,..)
In the case of restart request, isKillProcess is set false (in Renesas backend) to wait for GDB is exited normally.

bangdang1411 commented:
Hi @jonahgraham-san,
The latest commit fix both issue `Terminate `and `Restart `GDB server. It does not need to implement in `Renesas's backend`.
I see that cdt-gdb-adapter takes over 4 seconds to send `disconnect response` to vs code has no error returned.
I think the node process is not killed at this time and we can wait for the exiting gdb process completed.
```
[07:45:05.525 UTC] From client: disconnect({"restart":false})
[07:45:05.527 UTC] GDB command: 49 -gdb-exit
[07:45:05.543 UTC] GDB result: 49 exit
[07:45:05.543 UTC] GDB notify async: thread-group-exited,id="i1"
[07:45:08.535 UTC] To client: {"seq":0,"type":"event","event":"output","body":{"category":"server","output":"Disconnected from the Target Debugger.\r\n"}}
[07:45:08.551 UTC] To client: {"seq":0,"type":"event","event":"output","body":{"category":"server","output":"e2-server-gdb has exited with code 0"}}
[07:45:08.551 UTC] To client: {"seq":0,"type":"response","request_seq":10,"command":"disconnect","success":true}
```

jonahgraham commented on src/GDBTargetDebugSession.ts:
```diff
@@ -341,4 +347,45 @@ export class GDBTargetDebugSession extends GDBDebugSession {
+                if (os.platform() === 'win32') {
+                    this.gdbserver?.kill();
+                } else {
+                    this.gdbserver?.kill('SIGTERM');
```
Please check [documentation again](https://nodejs.org/api/child_process.html#subprocesskillsignal), on Linux SIGTERM is the default, that is why I was raising that as an issue as kill() on Windows is equivalent to SIGKILL, but kill() on Linux is SIGTERM.

jonahgraham reviewed (commented):
I need to spin this up as I want to understand why the code in VSCode is not being observed. That will take me another day.

bangdang1411 commented:
> I need to spin this up as I want to understand why the code in VSCode is not being observed. That will take me another day.

I try to change the code that does not send the response for` disconnect request`. In my environment, after 6 seconds VS Code will kill the node process (I assume the time is depended on the performance of PC). I think we need to extend the timeout on VS Code side. But this code can work for now and the change should be in VSCode repository.

> Please check [documentation again](https://nodejs.org/api/child_process.html#subprocesskillsignal), on Linux SIGTERM is the default, that is why I was raising that as an issue as kill() on Windows is equivalent to SIGKILL, but kill() on Linux is SIGTERM.

Sorry, I misunderstood your question. SIGTERM is only supported in Linux but it is can be ignored and the Promise that I implemented will wait forever. Is it correct? 
I changed both case Windows and Linux to 'SIGKILL'. My understanding is 'SIGKILL' can work in both environments and cannot be ignored.

jonahgraham commented:
> In my environment, after 6 seconds VS Code will kill the node process (I assume the time is depended on the performance of PC).

I have tested this on a couple of computers and it seems that node is killed in about 2 seconds on each of them.

However, I have a new proposal on how to resolve this problem. Instead of running cdt-gdb-adapter as a separate node process, we can run it in the extension host process. That means that the node process won't be killed at all when disconnect completes and we can do whatever we want. However this moves the problem in part to VSCode extension. Please look up DebugAdapterInlineImplementation in https://code.visualstudio.com/api/extension-guides/debugger-extension

In parallel, please raise an issue with vscode, perhaps with a PR, to remove the 2 second kill.

bangdang1411 commented:
Hi @jonahgraham-san,

> However, I have a new proposal on how to resolve this problem. Instead of running cdt-gdb-adapter as a separate node process, we can run it in the extension host process. That means that the node process won't be killed at all when disconnect completes and we can do whatever we want. However this moves the problem in part to VSCode extension. Please look up DebugAdapterInlineImplementation in https://code.visualstudio.com/api/extension-guides/debugger-extension

I understand your idea but it takes time to investigate and implement. Anyway, It should be implemented in Renesas vs code.
Is that correct? Additionally, when it was implemented, this PR should be merged to fix the issue gdb does not exit normally after 4 seconds.

> In parallel, please raise an issue with vscode, perhaps with a PR, to remove the 2 second kill.

I see the [same requirement](https://github.com/microsoft/vscode/issues/175850), but it is discussing in [another issue](https://github.com/microsoft/vscode/issues/172284)

Could you please share your idea? Thank you.

jonahgraham commented:
> Hi @jonahgraham-san,
> 
> > However, I have a new proposal on how to resolve this problem. Instead of running cdt-gdb-adapter as a separate node process, we can run it in the extension host process. That means that the node process won't be killed at all when disconnect completes and we can do whatever we want. However this moves the problem in part to VSCode extension. Please look up DebugAdapterInlineImplementation in https://code.visualstudio.com/api/extension-guides/debugger-extension
> 
> I understand your idea but it takes time to investigate and implement. Anyway, It should be implemented in Renesas vs code. Is that correct? 

The main change in cdt-gdb-vscode (and any extensions that use cdt-gdb-adapter too, line renesas-gdb-vscode). It needs review work up and down the entire stack because, for example, cdt-gdb-adapter was written assuming that it would be in its own process. For the most part it probably will work, but the entry points and cleanup especially will need some investigating to be sure.

> Additionally, when it was implemented, this PR should be merged to fix the issue gdb does not exit normally after 4 seconds.

I don't think so, at least not as is. Renesas can do a workaround for the issues that is affecting them in their own adapter. That is to say, this entire PR can just be done in Renesas's own code base as it is a workaround for Renesas's issue that does not work in the general case.

> 
> > In parallel, please raise an issue with vscode, perhaps with a PR, to remove the 2 second kill.
> 
> I see the [same requirement](https://github.com/microsoft/vscode/issues/175850), but it is discussing in [another issue](https://github.com/microsoft/vscode/issues/172284)
> 
> Could you please share your idea? Thank you.

I just read that and it is on point. Of course someone needs to provide a PR to improve the situation there based on what is discussed.

bangdang1411 commented:
> The main change in cdt-gdb-vscode (and any extensions that use cdt-gdb-adapter too, line renesas-gdb-vscode). It needs review work up and down the entire stack because, for example, cdt-gdb-adapter was written assuming that it would be in its own process. For the most part it probably will work, but the entry points and cleanup especially will need some investigating to be sure.

I am investigating the [VS Code Mock Debug](https://github.com/microsoft/vscode-mock-debug) and trying to know how it works. But it seems more difficult than my experience. Could you please help me take a look at it and give me some ideas on how to implement your solution? Thank you.

> I don't think so, at least not as is. Renesas can do a workaround for the issues that is affecting them in their own adapter. That is to say, this entire PR can just be done in Renesas's own code base as it is a workaround for Renesas's issue that does not work in the general case.

I understood. But please consider if we have a short-term solution for fixing the restart bug in Renesas. It should be fixed in the next release (no more time). Thank you.

> I just read that and it is on point. Of course, someone needs to provide a PR to improve the situation there based on what is discussed.

My understanding in that discussion is requested to improve the UX when the termination process is more than 2s. I think removing the 2-second kill is not acceptable.

jonahgraham commented:
I think the solution properly belongs in the Renesas extension so I am going to close this for now. If there is a generic version please provide it.

Please also see https://github.com/eclipse-cdt-cloud/cdt-gdb-adapter/issues/253#issuecomment-1480059599 for the next time you update to newer cdt-gdb-adapter.
